### ISSUE TITLE
[QUESTION] Stop Blazor View canvas from auto resizing

### ISSUE THREAD

**Issue by DanWBR** [mono/SkiaSharp#2038]
It seems that the canvas in Blazor is automatically resized to match the original proportions. I don't know if I'm missing something or if this is the default behavior. I would like to have the canvas to always fill the available width and height, in order to accommodate my application running in portrait and landscape orientations.

My CSS for the canvas is:

```
.canvas-container {
    line-height: 1;
}

    .canvas-container canvas {
        width: 100%;
        height: 100%;
        max-width: 100%;
        max-height: 100%;
    }
```

It works initially, but then when the window is resized, it keeps the original proportion and resizes itself accordingly.

**Comment by mgood7123** [comment]
you mean your content appears stretched when resizing?

**Comment by DanWBR** [comment]
@mgood7123 before you ask, the parent container is at 100% width and height, it is not preventing the canvas to fil the available space.

**Comment by mgood7123** [comment]
> @mgood7123 check this video, you'll notice that the width and height of the canvas are defined internally (I didn't set them explicitly in my code), and are updated automatically to keep the same initial aspect ratio. I'd like to have it at 100% width and height automatically.
> canvas.mov

if the aspect ratio is not kept then this WILL result in stretching

you could center it horizontally and vertically instead

**Comment by mgood7123** [comment]
maybe this?

![image](https://user-images.githubusercontent.com/26642847/168406972-ea31ae01-040c-44b0-8b6c-0c0b06d7f43c.png)

**Comment by DanWBR** [comment]
> > @mgood7123 check this video, you'll notice that the width and height of the canvas are defined internally (I didn't set them explicitly in my code), and are updated automatically to keep the same initial aspect ratio. I'd like to have it at 100% width and height automatically.
> 
> > canvas.mov
> 
> 
> 
> if the aspect ratio is not kept then this WILL result in stretching (assuming it is rendered in the same way as an image)
> 
> 
> 
> you could center it horizontally and vertically instead

@mgood7123 it should not stretch. What I need is the same behavior that a SkCanvasView Windows Forms control has when added to a form with its DockState set to Fill. The canvas keeps filling the form as you resize it.

**Comment by mgood7123** [comment]
> > > @mgood7123 check this video, you'll notice that the width and height of the canvas are defined internally (I didn't set them explicitly in my code), and are updated automatically to keep the same initial aspect ratio. I'd like to have it at 100% width and height automatically.
> > 
> > 
> > > canvas.mov
> > 
> > 
> > if the aspect ratio is not kept then this WILL result in stretching (assuming it is rendered in the same way as an image)
> > you could center it horizontally and vertically instead
> 
> @mgood7123 it should not stretch. What I need is the same behavior that a SkCanvasView Windows Forms control has when added to a form with its DockState set to Fill. The canvas keeps filling the form as you resize it.

not sure how else to help

**Comment by mattleibow** [comment]
I believe this is the way the canvas works.

The CSS is used to set the desired size, and then in order to make sure the image is not blurry, the html size is set to the CSS size * density:

https://developer.mozilla.org/en-US/docs/Web/API/Window/devicePixelRatio#correcting_resolution_in_a_canvas

**Comment by DanWBR** [comment]
@mattleibow @mgood7123 the following CSS did the trick:

```
.canvas-container {
    width: 100%;
    height: 100%;
    max-width: 100%;
    max-height: 100%;
}

    .canvas-container canvas {
        position: absolute;
        top: 0;
        bottom: 0;
        left: 0;
        right: 0;
    }
```

<img width="1369" alt="Captura de Tela 2022-05-14 às 08 38 03" src="https://user-images.githubusercontent.com/2669419/168426101-440b5dbf-c40b-472a-89f8-9053642bf6fe.png">

I'm closing this, thank you very much for your help!

**Comment by DanWBR** [comment]
reopening. it worked but then the canvas stayed at the top z-order blocking all other elements on the page.

**Comment by mattleibow** [comment]
I think this is just your CSS. I have a page here with all the heights set to 100% and then the canvas fits the screen without worrying about aspect.

I am not sure how to actually get what you want. 

Also, are you sure you don't have other CSS/JS that is resizing the canvas?


https://user-images.githubusercontent.com/1096616/168792722-45e9bba6-5928-404c-bd5e-c21fd8c9eed1.mp4

**Comment by DanWBR** [comment]
@mattleibow it suddenly started working as expected. I'm still learning to work with CSS and JS, so yes, it was on me. Thank you very much!

BTW, the blazor GL view is not working well on iOS. Almost nothing is drawn.

![image](https://user-images.githubusercontent.com/2669419/168793503-e92fb335-8013-4a2c-b0f4-0dd425083044.png)

**Comment by mattleibow** [comment]
Oh dear. Could you open an issue for the iOS issue?

I see things there so maybe also let me know what is drawn and what is not. Also, if you can get a smaller repro it would be great.

And, does it work with the SKCanvasView? I wonder if the fact the lines are not working it is something to do with my stencil buffers not being right...


Or maybe iOS is finally blowing up on GL and demanding Metal usage. Apple is threatening us... So yeah, what version of iOS is this?

**Comment by DanWBR** [comment]
@mattleibow yes, it works nicely with SKCanvasView. I'll create an issue with a comparison between the two. Thanks

I'll close this as I've managed to get it working with CSS stuff only. Height is not 100% in order to prevent the scrollbar from showing up, and then having the touchmove event working as expected (dragging objects), not trying to scroll the canvas up and down.

```
.canvas-container {
    width: 100%;
    height: 90%;
    max-width: 100%;
    max-height: 90%;
}

    .canvas-container canvas {
        width: 100%;
        height: 98%;
        max-width: 100%;
        max-height: 98%;
        touch-action: none;
    }
```
